v2/cmd/asoctl/internal/importing/importable_arm_resource.go

```diff
@@ -556,6 +560,48 @@ func (i *importableARMResource) SetOwner(
 	}
 }
 
+var safeResourceNameMappings = map[rune]rune{
+	'_':  '-', // underscores are replaced with hyphens
+	'/':  '-', // slashes are replaced with hyphens
+	'\\': '-', // backslashes are replaced with hyphens
+	'%':  '-', // percent signs are replaced with hyphens
+}
+
+var safeResourceNameRegex = regexp.MustCompile("-+")
+
+// safeResourceName ensures the name is a valid Kubernetes name
+func safeResourceName(name string) string {
+	buffer := make([]rune, 0, len(name))
+
+	for _, r := range name {
+		if unicode.IsLetter(r) || unicode.IsNumber(r) {
```

Comment:
Would crafting this like this be clearer?
```
	for _, r := range name {
		if unicode.IsLetter(r) || unicode.IsNumber(r) {
			buffer = append(buffer, unicode.ToLower(r))
		} else if len(buffer) == 0 { 
			// Discard leading special characters so that the result always starts with a letter or number
			continue
		} else if c, ok := safeResourceNameMappings[r]; ok {
			buffer = append(buffer, c)
		} else if unicode.IsSpace(r) {
			buffer = append(buffer, '-')
		}
	}
```

For whatever reason it took me a second when reading this to realize "oh all of these cases are mutually exclusive"

Reply:
I don't find a chain of **if**..**else** to be very readable; instead, I refactored it to use a **switch**, which I think is clearer than both what I had and what you suggested.